sklearn/utils/estimator_checks.py

```diff
@@ -4133,11 +4133,79 @@ def check_param_validation(name, estimator_orig):
-                    if any(
-                        X_type.endswith("labels")
-                        for X_type in _safe_tags(estimator, key="X_types")
-                    ):
-                        # The estimator is a label transformer and take only `y`
-                        getattr(estimator, method)(y)
-                    else:
-                        getattr(estimator, method)(X, y)
+                    getattr(estimator, method)(X, y)
+
+
+def check_set_output(name, transformer_orig):
+    # Check transformer.set_output configures the output of `transform`
+    tags = transformer_orig._get_tags()
+    if "2darray" not in tags["X_types"] or tags["no_validation"]:
+        return
+
+    rng = np.random.RandomState(0)
+    transformer = clone(transformer_orig)
+
+    X = rng.uniform(size=(20, 5))
+    X = _pairwise_estimator_convert_X(X, transformer_orig)
+    y = rng.randint(0, 2, size=20)
+    y = _enforce_estimator_tags_y(transformer_orig, y)
+    set_random_state(transformer)
+
+    X_trans_no_setting = transformer.fit_transform(X, y)
+    if name in CROSS_DECOMPOSITION:
+        X_trans_no_setting = X_trans_no_setting[0]
+
+    transformer.set_output(transform="default")
+    X_trans_default = transformer.fit_transform(X, y)
+    if name in CROSS_DECOMPOSITION:
+        X_trans_default = X_trans_default[0]
+
+    # Default and no setting returne the same transform
+    assert_allclose_dense_sparse(X_trans_no_setting, X_trans_default)
+
+
+def check_set_output_pandas(name, transformer_orig):
+    # Check transformer.set_output configures the output of transform="pandas"
+    try:
+        import pandas as pd
+    except ImportError:
+        raise SkipTest(
+            "pandas is not installed: not checking column name consistency for pandas"
+        )
+
+    tags = transformer_orig._get_tags()
+    if "2darray" not in tags["X_types"] or tags["no_validation"]:
+        return
+
+    rng = np.random.RandomState(0)
+    transformer = clone(transformer_orig)
+
+    X = rng.uniform(size=(20, 5))
+    X = _pairwise_estimator_convert_X(X, transformer_orig)
+    y = rng.randint(0, 2, size=20)
+    y = _enforce_estimator_tags_y(transformer_orig, y)
+    set_random_state(transformer)
+
+    feature_names_in = [f"col{i}" for i in range(X.shape[1])]
+    df = pd.DataFrame(X, columns=feature_names_in)
+
+    X_trans_no_setting = transformer.fit_transform(df, y)
+    if name in CROSS_DECOMPOSITION:
+        X_trans_no_setting = X_trans_no_setting[0]
+    transformer.set_output(transform="pandas")
+
+    try:
+        X_trans_pandas = transformer.fit_transform(df, y)
+    except ValueError as e:
+        # transformer does not support sparse data
+        assert str(e) == "Pandas output does not support sparse data", e
+        return
+    if name in CROSS_DECOMPOSITION:
+        X_trans_pandas = X_trans_pandas[0]
+
+    # Return is a pandas DataFrame
+    assert hasattr(X_trans_pandas, "iloc")
```

Comment:
since the test is specialized for pandas and we already imported pandas, we could check that this is definetely a pandas dataframe.

Reply:
When I see the assertions below, I think that we could even use https://pandas.pydata.org/docs/reference/api/pandas.testing.assert_frame_equal.html to check all of those in one line.